# ¿Qué es un algoritmo?

En este notebook, empezaremos nuestro estudio de algoritmos computacionales.

Un **algoritmo** es una "receta" computacional, que consiste en una serie de instrucciones para que la computadora lleve a cabo un cálculo dado. Gran parte del curso consistirá en desarrollar algoritmos para calcular, de forma numérica, distintas cantidades en la física, a partir de algún modelo matemático. El campo que se ocupa de diseñar y estudiar estos algoritmos es el **análisis numérico**. Su aplicación a problemas de física se puede decir que constituye la **física computacional**.

Algunos algoritmos (por ejemplo, la eliminación gaussiana que veremos más adelante) proveen una manera de llevar a cabo un cálculo de manera "exacta" (dentro de las restricciones impuestas por el uso de números con precisión finita) en un número finito de pasos.

Sin embargo, en general, no podemos esperar que haya una fórmula analítica cerrada para calcular las cantidades de interés de manera exacta. En este caso, será necesario emplear un algoritmo **iterativo**, que en principio podría correr ¡por un tiempo infinito! Lo pararemos cuando pensemos que el problema ya se resolvió de forma "suficientemente buena".

## Algoritmos iterativos

Un **algoritmo iterativo** repite un mismo cálculo un gran número de veces, modificando un valor (o varios valores) en el proceso, hasta que (en el mejor de los casos) converja a una solución.

Un algorithmo iterativo empieza desde una adivinanza inicial $x_0$, y aplica un procedimiento / receta matemática, o sea alguna función $f$ (que puede ser complicada), para producir una siguiente adivinanza $x_1 := f(x_0)$. Esto se repite para producir una secuencia $x_0, x_1, \ldots, x_n, \ldots$, tales que

$$x_1 = f(x_0)$$
$$x_2 = f(x_1)$$
$$x_3 = f(x_2)$$

etc. En general, escribimos una iteración como 

$$x_{n+1} := f(x_n).$$

La esperanza es que la secuencia $x_n$ converja hacia un valor límite $x^*$ cuando $n \to \infty$, y que el $x^*$ resultante sea solución del problema original.

Dado que no podemos llevar a cabo la iteración un número infinito de veces, se corta la iteración después de un cierto número de pasos, para dar una solución *aproximada*, que se acerca dentro de cierta *tolerancia* al resultado teórico exacto $x^*$. Por lo tanto, cualquier algoritmo iterativo requiere una condición de terminación.

En la computadora, no pensamos en escribir $x_n$, sino pensamos en el **valor actual** de $x$, y el **valor siguiente** de $x$. Dentro del bucle, usamos el valor actual de $x$ para calcular el valor nuevo. Al final del bucle, debemos actualizar el "valor actual".

# Iteraciones de punto fijo

**[1]** (i) Define la función $f_1(x) = \frac{x}{2} - 1$.

(ii) Toma una condición inicial $x_0$ y lleva a cabo la iteración a mano. ¿Qué observas?

(iii) Utiliza un bucle `for` para ver cómo son los primeros $N$ iterados $x_n$. Haz una función que tome como argumento $x_0$ y $N$.

(iv) Repite la iteración para varios valores de $x_0$. ¿Qué observas?

In [1]:
f(x) = x/2 - 1
x0 = 5

5

In [2]:
x0 = f(x0)

1.5

In [3]:
x0 = f(x0)

-0.25

In [4]:
x0 = f(x0)

-1.125

In [5]:
x0 = f(x0)

-1.5625

In [6]:
x0 = f(x0)

-1.78125

In [7]:
x0 = f(x0)

-1.890625

In [8]:
x0 = f(x0)

-1.9453125

In [9]:
x0 = f(x0)

-1.97265625

In [10]:
x0 = f(x0)

-1.986328125

In [11]:
x0 = f(x0)

-1.9931640625

In [12]:
x0 = 5
for i in 1:20
    x0 = f(x0)
    @show x0
end

x0 = 1.5
x0 = -0.25
x0 = -1.125
x0 = -1.5625
x0 = -1.78125
x0 = -1.890625
x0 = -1.9453125
x0 = -1.97265625
x0 = -1.986328125
x0 = -1.9931640625
x0 = -1.99658203125
x0 = -1.998291015625
x0 = -1.9991455078125
x0 = -1.99957275390625
x0 = -1.999786376953125
x0 = -1.9998931884765625
x0 = -1.9999465942382812
x0 = -1.9999732971191406
x0 = -1.9999866485595703
x0 = -1.9999933242797852


In [13]:
function iterar1(x0, n)
    xi = x0
    @show xi
    for i in 1:n
        xi = f(xi)
        @show xi
    end
end

iterar1 (generic function with 1 method)

In [14]:
iterar1(10, 20)

xi = 10
xi = 4.0
xi = 1.0
xi = -0.5
xi = -1.25
xi = -1.625
xi = -1.8125
xi = -1.90625
xi = -1.953125
xi = -1.9765625
xi = -1.98828125
xi = -1.994140625
xi = -1.9970703125
xi = -1.99853515625
xi = -1.999267578125
xi = -1.9996337890625
xi = -1.99981689453125
xi = -1.999908447265625
xi = -1.9999542236328125
xi = -1.9999771118164062
xi = -1.9999885559082031


In [15]:
iterar1(20,20)

xi = 20
xi = 9.0
xi = 3.5
xi = 0.75
xi = -0.625
xi = -1.3125
xi = -1.65625
xi = -1.828125
xi = -1.9140625
xi = -1.95703125
xi = -1.978515625
xi = -1.9892578125
xi = -1.99462890625
xi = -1.997314453125
xi = -1.9986572265625
xi = -1.99932861328125
xi = -1.999664306640625
xi = -1.9998321533203125
xi = -1.9999160766601562
xi = -1.9999580383300781
xi = -1.999979019165039


In [16]:
iterar1(30,20)

xi = 30
xi = 14.0
xi = 6.0
xi = 2.0
xi = 0.0
xi = -1.0
xi = -1.5
xi = -1.75
xi = -1.875
xi = -1.9375
xi = -1.96875
xi = -1.984375
xi = -1.9921875
xi = -1.99609375
xi = -1.998046875
xi = -1.9990234375
xi = -1.99951171875
xi = -1.999755859375
xi = -1.9998779296875
xi = -1.99993896484375
xi = -1.999969482421875


In [17]:
iterar1(40,20)

xi = 40
xi = 19.0
xi = 8.5
xi = 3.25
xi = 0.625
xi = -0.6875
xi = -1.34375
xi = -1.671875
xi = -1.8359375
xi = -1.91796875
xi = -1.958984375
xi = -1.9794921875
xi = -1.98974609375
xi = -1.994873046875
xi = -1.9974365234375
xi = -1.99871826171875
xi = -1.999359130859375
xi = -1.9996795654296875
xi = -1.9998397827148438
xi = -1.9999198913574219
xi = -1.999959945678711


Para distintos valores de $x_0$ tras algunas iteraciones el valor de $x_i$ siempre se va hacía el mismo punto.

**[2]** (i) Define una función `iterar` que lleva a cabo lo que hiciste en la pregunta 1. Debe aceptar como su primer argumento (el nombre de) *la función `f` que iterar*, así como el número de veces que se iterará, y la condición inicial.

(ii) Utilízalo para iterar la función $f_2(x) = \cos(x)$.

(iii) ¿Adónde converge la iteración? ¿Cuál ecuación hemos resuelto?


In [18]:
function iterar2(f, n, x0)
    xi = x0
    @show xi
    for i in 1:n
        xi = f(xi)
        @show xi
    end
end

iterar2 (generic function with 1 method)

In [19]:
iterar2(cos, 10, 5)

xi = 5
xi = 0.28366218546322625
xi = 0.9600369302946615
xi = 0.5734897326953653
xi = 0.8400126809521591
xi = 0.6674533830038623
xi = 0.7854005359989481
xi = 0.7071051035019478
xi = 0.7602456869604484
xi = 0.7246667298504657
xi = 0.7487203836426739


La ecuación converge hacia $x\approx0.7487$, se está resolviendo la ecuación $\cos (x) = x$.

**[3]** Ahora queremos graficar. 

(i) Modifica la función `iterar` para que vaya guardando los valores de $x_n$ en un arreglo, el cual regresa.

(ii) Utiliza el paquete `Plots` para graficar el resultado. ¡Ten cuidado con el tipo de gráfica que dibujas: deben ser puntos! Para eso puedes utilizar la función `scatter`.

(iii) Grafica la trayectoria para varios valores de $x_0$ en una sola gráfica. [Utiliza `scatter!`, con `!` al final, para *agregar* más información a un `plot` ya existente.] ¿Qué observas?

(iv) Importa el paquete `Interact` y utiliza `@manipulate` antes de un bucle `for` sobre `x_0` para ver cómo cambia la visualización (de una sola condición inicial) de forma interactiva.

In [1]:
function iterar3(f, n, x0)
    xi = x0
    iteraciones = []
    for i in 1:n
        xi = f(xi)
        push!(iteraciones, xi)
    end
    return iteraciones
end

iterar3 (generic function with 1 method)

In [2]:
iterar3(cos, 20, 10)

20-element Array{Any,1}:
 -0.839072
  0.668154
  0.784967
  0.707412
  0.760046
  0.724804
  0.748629
  0.732622
  0.743423
  0.736156
  0.741055
  0.737757
  0.739979
  0.738483
  0.739491
  0.738812
  0.739269
  0.738961
  0.739169
  0.739029

In [8]:
using Plots
using Interact

scatter(iterar3(cos, 20, 10))
scatter!(iterar3(cos, 20, 15))
scatter!(iterar3(cos,20, 20))
scatter!(iterar3(cos, 20, 25))

In [13]:
gr()
Pkg.add("Gadfly")

INFO: Cloning cache of AxisAlgorithms from https://github.com/timholy/AxisAlgorithms.jl.git
INFO: Cloning cache of Calculus from https://github.com/johnmyleswhite/Calculus.jl.git
INFO: Cloning cache of CommonSubexpressions from https://github.com/rdeits/CommonSubexpressions.jl.git
INFO: Cloning cache of Compose from https://github.com/GiovineItalia/Compose.jl.git
INFO: Cloning cache of CoupledFields from https://github.com/Mattriks/CoupledFields.jl.git
INFO: Cloning cache of DataArrays from https://github.com/JuliaStats/DataArrays.jl.git
INFO: Cloning cache of DataFrames from https://github.com/JuliaStats/DataFrames.jl.git
INFO: Cloning cache of DiffBase from https://github.com/JuliaDiff/DiffBase.jl.git
INFO: Cloning cache of Distances from https://github.com/JuliaStats/Distances.jl.git
INFO: Cloning cache of Distributions from https://github.com/JuliaStats/Distributions.jl.git
INFO: Cloning cache of DualNumbers from https://github.com/JuliaDiff/DualNumbers.jl.git
INFO: Cloning cache o

In [14]:
using Gadfly

INFO: Precompiling module Gadfly.
This may mean module Measures does not support precompilation but is imported by a module that does.
ERROR: LoadError: Declaring __precompile__(false) is not allowed in files that are being precompiled.
Stacktrace:
 [1] _require(::Symbol) at ./loading.jl:448
 [2] require(::Symbol) at ./loading.jl:398
 [3] include_from_node1(::String) at ./loading.jl:569
 [4] include(::String) at ./sysimg.jl:14
 [5] anonymous at ./<missing>:2
while loading /Users/uzielnmtz/.julia/v0.6/Compose/src/Compose.jl, in expression starting on line 9
ERROR: LoadError: Failed to precompile Compose to /Users/uzielnmtz/.julia/lib/v0.6/Compose.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:703
 [2] _require(::Symbol) at ./loading.jl:456
 [3] require(::Symbol) at ./loading.jl:398
 [4] include_from_node1(::String) at ./loading.jl:569
 [5] include(::String) at ./sysimg.jl:14
 [6] anonymous at ./<missing>:2
while loading /Users/uzielnmtz/.julia/v0.6/Gadfly/src/Gadfly.jl, in 

LoadError: Failed to precompile Gadfly to /Users/uzielnmtz/.julia/lib/v0.6/Gadfly.ji.

In [10]:
@manipulate for x0 = -10:2:10
    x = 0:20
    scatter(iterar3(cos, 20, x0))
end

Interact.Options{:SelectionSlider,Int64}(13: "input-4" = 0 Int64 , "x0", 0, "0", Interact.OptionDict(DataStructures.OrderedDict("-10"=>-10,"-8"=>-8,"-6"=>-6,"-4"=>-4,"-2"=>-2,"0"=>0,"2"=>2,"4"=>4,"6"=>6,"8"=>8…), Dict(2=>"2",0=>"0",-4=>"-4",10=>"10",8=>"8",6=>"6",-6=>"-6",4=>"4",-8=>"-8",-10=>"-10"…)), Any[], Any[], true, "horizontal")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 5 
 
 
 10 
 
 
 15 
 
 
 20 
 
 
 0.6 
 
 
 0.7 
 
 
 0.8 
 
 
 0.9 
 
 
 1.0 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 y1

**[4]** ¿Qué ocurre si iteras la función $f_3(x) = 2x + 1$?

In [24]:
@manipulate for x0 = -10:2:10
    scatter(iterar3(x -> 2*x + 1, 30, x0))
end

Interact.Options{:SelectionSlider,Int64}(5: "input-2" = 0 Int64 , "x0", 0, "0", Interact.OptionDict(DataStructures.OrderedDict("-10"=>-10,"-8"=>-8,"-6"=>-6,"-4"=>-4,"-2"=>-2,"0"=>0,"2"=>2,"4"=>4,"6"=>6,"8"=>8…), Dict(2=>"2",0=>"0",-4=>"-4",10=>"10",8=>"8",6=>"6",-6=>"-6",4=>"4",-8=>"-8",-10=>"-10"…)), Any[], Any[], true, "horizontal")

La función no converge, para distintos valores de $x_0$ ninguna iteración converge.

**[5]** (i) Pensando en la ecuación $x_{n+1} = f(x_n)$, en el límite cuando $n \to \infty$, si es que la iteración converge a un valor que podemos llamar $x_\infty$, ¿a cuál valor debe converger? [Pista: toma el límite de los dos lados de la ecuación.] ¿Coincide con lo observado para $f_1$ y $f_2$?

(ii) Graficamente, ¿a qué corresponde resolver la ecuación para $x_\infty$? Dibuja las gráficas correspondientes para $f_1$ y $f_2$ y checa tu respuesta. 

(iii) Puedes adivinar cuál es la condición para que la iteración converja?

$$\lim_{n \to \infty} x_{n+1} = \lim_{n \to \infty} f(x_n) \\\\
\Rightarrow x_\infty = f(x_\infty)$$

De manera que el problema que estamos intentando resolver es el de encontrar $x$ tal que $f(x) = x$ para alguna función.

Para nuestros dos ejemplos en particular tenemos:

$$\cos(x) = x \Rightarrow x\approx 0.739029 \\\\
\dfrac{x}{2} - 1 = x \Rightarrow x\approx -1.999959$$

In [25]:
plot(x -> x / 2 - 1)
plot!(cos)
plot!(x -> x)

Gráficamente el problema a resolver es el de las intersecciónes de las funciones $f_1$ y $f_2$ con $g(x)=x$.

Tal vez una condición para que converja es es que no haya divergencias de la función en las cercanías del guess y que el valor deseado no se encuentre muy lejos.

**[6]** A menudo, se puede utilizar una iteración de este tipo para resolver ecuaciones. 

(i) Inventa una iteración de la forma $x_{n+1} = f(x_n)$ para resolver la ecuación $x^2 + x - 1 = 0$. ¿Para cuáles $x_0$ funciona? ¿A cuál solución converge?

(ii) Inventa otra. ¿Funciona para $x_0$ diferentes?

(iii) Nota que hay algunas iteraciones que **no converjan**. Por ejemplo, ¿qué ocurre con la iteración $x_{n+1} = 1 - x_n^2$?

In [26]:
@manipulate for x0=-20:2:20
    scatter(iterar3(x -> (x + 1/x) / 2 - 1/2, 40, x0))
end

Interact.Options{:SelectionSlider,Int64}(9: "input-3" = 0 Int64 , "x0", 0, "0", Interact.OptionDict(DataStructures.OrderedDict("-20"=>-20,"-18"=>-18,"-16"=>-16,"-14"=>-14,"-12"=>-12,"-10"=>-10,"-8"=>-8,"-6"=>-6,"-4"=>-4,"-2"=>-2…), Dict(-16=>"-16",2=>"2",16=>"16",0=>"0",-20=>"-20",-4=>"-4",10=>"10",8=>"8",6=>"6",-6=>"-6"…)), Any[], Any[], true, "horizontal")

En este caso la iteración no converge para $x_0 = 0$ debido a una indeterminación en la iteración.

Dependiendo de donde esté el "guess" converge a la raíz positiva o negativa de la ecuación.

In [28]:
@manipulate for x0=-20:2:20
    scatter(iterar3(x -> (x - x^2) / 2 + 1 / 2, 40, x0))
end

Interact.Options{:SelectionSlider,Int64}(15: "input-5" = 0 Int64 , "x0", 0, "0", Interact.OptionDict(DataStructures.OrderedDict("-20"=>-20,"-18"=>-18,"-16"=>-16,"-14"=>-14,"-12"=>-12,"-10"=>-10,"-8"=>-8,"-6"=>-6,"-4"=>-4,"-2"=>-2…), Dict(-16=>"-16",2=>"2",16=>"16",0=>"0",-20=>"-20",-4=>"-4",10=>"10",8=>"8",6=>"6",-6=>"-6"…)), Any[], Any[], true, "horizontal")

Para esta iteración sólo converge para $x_0$ entre 0 y 2.

In [29]:
@manipulate for x0 = -20:2:20
    scatter(iterar3(x -> 1 - x^2, 40, x0))
end

Interact.Options{:SelectionSlider,Int64}(19: "input-6" = 0 Int64 , "x0", 0, "0", Interact.OptionDict(DataStructures.OrderedDict("-20"=>-20,"-18"=>-18,"-16"=>-16,"-14"=>-14,"-12"=>-12,"-10"=>-10,"-8"=>-8,"-6"=>-6,"-4"=>-4,"-2"=>-2…), Dict(-16=>"-16",2=>"2",16=>"16",0=>"0",-20=>"-20",-4=>"-4",10=>"10",8=>"8",6=>"6",-6=>"-6"…)), Any[], Any[], true, "horizontal")

## Bucles `while`

En lo anterior, usamos un bucle `for`, que requiere que sepamos de antemano el número de iteraciones que queramos.
Sin embargo, en este tipo de problemas, es más natural esperar **hasta que** converja, sin saber cuánto tiempo tomará.

Para esto, podemos ocupar otro tipo de bucle, un bucle `while` ("mientras", en español). Un bucle de este tipo repite los comandos en el cuerpo del bucle **mientras** una condición siga cierta. Su sintaxis es como sigue:

```
while <condicion>
    [haz esto]
    [y esto]
end
```

Sin embargo, para evitar bucles infinitos, a menudo es sensato incluir un contador para que no pueda haber demasiadas (posiblemente infinitas) iteraciones.

Mientras que en un bucle `for` hay un contador que se actualiza automáticamente, en un bucle `while` **nosotros somos los responsables** de tener una variable que actúe como contador.

**[7]** (i) Utilice un bucle `while` para contar de 1 a 10.

(ii) Utilice un bucle `while` para encontrar la potencia de 2 más grande debajo de un número dado.

(iii) Repite lo mismo con un bucle `for`, usando la palabra clave `break` para salir del bucle cuando una cierta condición se satisfaga.

In [30]:
i = 0
while i <= 10
    @show i
    i += 1
end

i = 0
i = 1
i = 2
i = 3
i = 4
i = 5
i = 6
i = 7
i = 8
i = 9
i = 10


In [31]:
n = 10
i = 0
potencia = 1
while i <= n
    if isinteger(sqrt(i))
        potencia = i
    end
    i += 1
end
@show potencia

potencia = 9


9

In [32]:
potencia = 1
for i in 1:20
    if isinteger(sqrt(i))
        potencia = i
    end
    if i == 10
        break
    end
end
@show potencia
    

potencia = 9


9

## Bucles while e iteraciones de punto fijo

**[8]** Utiliza un bucle `while` para la iteración de la función $f_1$. Fija una **tolerancia** razonable, y repite la iteración **hasta que** la distancia entre un iterado y el siguiente sea menor a la tolerancia. [Pista: ¿Cuál función matemática da la distancia entre dos números en una dimensión. Encuentra la función de Julia que lo hace.] 

**[9]** De la misma forma, escribe una versión nueva de la función `iterar`, que utiliza un `while` y acepta una tolerancia como argumento.

In [33]:
ϵ = 0.0000000000000000001
err = 1
f1(x) = x/2 - 1
x0 = 1
n = 0
while ϵ <= err
    x = x0
    x0 = f1(x0)
    err = norm(x-x0)
    n += 1
end
@show x0, n
    
    

(x0, n) = (-2.0, 56)


(-2.0, 56)

In [34]:
function iterar(f, ϵ, x0)
    err = 1
    n = 0
    iters = Float64[]
    numiters= []
    while ϵ <= err
        n += 1
        x = x0
        x0 = f(x0)
        err = norm(x-x0)
        push!(iters, x)
        push!(numiters, n)
    end
    return [numiters iters]
end

iterar (generic function with 1 method)

## Haz tu propia biblioteca 

Guarda la función `iterar` en un archivo que se llama `herramientas.jl`. Iremos agregando más métodos a este archivo.

Se incluye en un notebook o un script de Julia con `include("herramientas.jl")`.